### PR TITLE
Add LDTs to ACVP Tool

### PR DESCRIPTION
**Pull request by billbo-yang**
### Issues:
Resolves CryptoAlg-1862

### Description of changes: 
AWS-LC's ACVP tool currently doesn't support large data tests (LDTs), which are a new type of test for hash algorithms introduced for new validations. Since we're doing a new FIPS validation for AWS-LC, we now need to support these tests in order to produce ACVP response vectors.

### Call-outs:
There's a MacOS x86 FIPS GitHub CI failure that isn't happening on the other CI environments. Opening as a draft to assist in debugging the issue.

### Testing:
Tested by running on vectors (SHA-1, SHA2-224, SHA2-256, SHA2-384, SHA2-512, SHA2-512/256, SHA3-224, SHA3-256, SHA3-384, SHA3-512) with LDTs and ensuring that the output matches what is expected.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and the ISC license.

### CODE REVIEW

File: util/fipstools/acvp/modulewrapper/modulewrapper.cc [hunk 1]
```diff
@@ -1086,6 +1098,46 @@ static bool HashMCTSha3(const Span<const uint8_t> args[],
       {Span<const uint8_t>(md[1000])});
 }
 
+static unsigned char* BuildLDTMessage(const bssl::Span<const uint8_t> part_msg, int times) {
+  size_t full_msg_size = part_msg.size() * times;
+  unsigned char* full_msg = (unsigned char*) malloc (full_msg_size);
+  for(int i = 0; i < times; i++) {
+    memcpy(full_msg + i * part_msg.size(), part_msg.data(), part_msg.size());
```
Comment (andrewhop) [inline util/fipstools/acvp/modulewrapper/modulewrapper.cc:1105]:
```suggestion
    OPENSSL_memcpy(full_msg + i * part_msg.size(), part_msg.data(), part_msg.size());
```
And a few spots down below.

File: util/fipstools/acvp/modulewrapper/modulewrapper.cc [hunk 2]
```diff
@@ -1086,6 +1098,46 @@ static bool HashMCTSha3(const Span<const uint8_t> args[],
       {Span<const uint8_t>(md[1000])});
 }
 
+static unsigned char* BuildLDTMessage(const bssl::Span<const uint8_t> part_msg, int times) {
```
Comment (andrewhop) [inline util/fipstools/acvp/modulewrapper/modulewrapper.cc:1101]: Do you have a link to where this test is defined and how the message should be constructed?

File: CMakeLists.txt [hunk 3]
```diff
@@ -912,21 +912,38 @@ if(BUILD_TESTING)
...
+        list(APPEND acvp_sha_test_commands
+              COMMAND ${GO_EXECUTABLE} run check_expected.go
+                        -tool ${ACVP_TOOL}
+                        -module-wrappers modulewrapper:$<TARGET_FILE:modulewrapper>,testmodulewrapper:${TEST_WRAPPER}
+                        -tests "sha-tests/${test_name}")
+      endforeach()
+
       add_custom_target(
-        acvp_tests
+        acvp_go_install
```
Comment (andrewhop) [inline CMakeLists.txt:927]: NP: this feels more like build_acvp_tool, it's not really installing anything.

File: util/fipstools/acvp/modulewrapper/modulewrapper.cc (hunk 1 above)
Comment (billbo-yang) [reply util/fipstools/acvp/modulewrapper/modulewrapper.cc:1105]: Interestingly, the rest of `modulewrapper.cc` uses the regular `memcpy()`, so I used it here as well to follow precedent. I could switch it here and open a new PR to switch the rest to use `OPENSSL_memcpy()` if you think it's worth it though

Comment (andrewhop) [reply util/fipstools/acvp/modulewrapper/modulewrapper.cc:1108]: Nah, that would just create more merge conflicts in the future.

File: util/fipstools/acvp/modulewrapper/modulewrapper.cc [hunk 4]
```diff
@@ -1086,6 +1098,49 @@ static bool HashMCTSha3(const Span<const uint8_t> args[],
...
+  uint8_t digest[DigestLength];
+  const EVP_MD *md = MDFunc();
+  unsigned int md_out_size = DigestLength;
+
+  int times;
+  memcpy(&times, args[1].data(), sizeof(int));
+
+  unsigned char *msg = BuildLDTMessage(args[0], times);
+
+  EVP_Digest(msg, args[0].size() * times, digest, &md_out_size, md, NULL);
```
Comment (samuel40791765) [inline util/fipstools/acvp/modulewrapper/modulewrapper.cc:1139]: NP: in C++ land
```suggestion
  EVP_Digest(msg, args[0].size() * times, digest, &md_out_size, md, nullptr);
```

File: util/fipstools/acvp/modulewrapper/modulewrapper.cc [hunk 5]
```diff
@@ -1086,6 +1098,49 @@ static bool HashMCTSha3(const Span<const uint8_t> args[],
...
+template <const EVP_MD *(MDFunc)(), size_t DigestLength>
+static bool HashLDTSha3(const Span<const uint8_t> args[], ReplyCallback write_reply) {
+  uint8_t digest[DigestLength];
+  const EVP_MD *md = MDFunc();
+  unsigned int md_out_size = DigestLength;
+
+  int times;
+  memcpy(&times, args[1].data(), sizeof(int));
+
+  unsigned char *msg = BuildLDTMessage(args[0], times);
```
Comment (samuel40791765) [inline util/fipstools/acvp/modulewrapper/modulewrapper.cc:1137]: Is it possible to use `std::unique_ptr` to manage the memory freeing for us?

**Review by andrewhop — APPROVED** [review]

**Review by samuel40791765 — APPROVED** [review]

**Review by andrewhop — APPROVED** [review]

**Pull request merged**